File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.py
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        
        last_day = days[-1]
        dp = [inf] * (last_day + 1)
        dp[0] = 0
        days = set(days)
        
        for i in range(1, last_day+1):
            if i not in days:
                dp[i] = dp[i-1]
                continue
            
            dp[i] = min(dp[i], costs[0] + dp[max(0, i-1)]) 
            dp[i] = min(dp[i], costs[1] + dp[max(0, i-7)])
            dp[i] = min(dp[i], costs[2] + dp[max(0, i-30)])
        
        return dp[last_day]
        

File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.kt
class Solution {
    fun mincostTickets(days: IntArray, costs: IntArray): Int {
        val dp = IntArray(366){1000 * 366}
        dp[0] = 0
        for (day in 1..365) {
            if (!days.contains(day)) {
                dp[day] = dp[day-1]
            } else {
                dp[day] = Math.min(dp[day], costs[0] + dp[Math.max(0, day-1)])
                dp[day] = Math.min(dp[day], costs[1] + dp[Math.max(0, day-7)])
                dp[day] = Math.min(dp[day], costs[2] + dp[Math.max(0, day-30)])
            }
        }
        return dp[365]
    }
}